api/jobsendpoint.go

```diff
@@ -0,0 +1,100 @@
+package api
+
+import (
+	"fmt"
+	"github.com/cyverse-de/de-stats/cron"
+	"github.com/cyverse-de/de-stats/util"
+	"github.com/labstack/echo"
+	"net/http"
+	"time"
+)
+
+type JobsParams struct {
+	//The beginning of the time period of the response
+	//
+	//in: query
+	//required: false
+	//default: one week ago
+	StartDate string
+
+	//The end of the time period of the response
+	//
+	//in: query
+	//required: false
+	//default: today
+	EndDate string
+}
+
+type JobsResponse struct {
+	Count int	 	`json:"count"`
+	Jobs []cron.Job `json:"jobs"`
+}
+
+func verifyJobsParameters(ctx echo.Context) (string, string, error) {
+	const (
+		dateFormat = "20060102"
+	)
+
+	currentTime := time.Now()
+	oneWeekAgo := time.Now().AddDate(0, 0, -7)
+
+	startDate, err := util.StringQueryParam(ctx, "startDate", oneWeekAgo.Format(dateFormat))
+	fmt.Println(startDate)
+	if err != nil {
+		return "", "", ctx.JSON(http.StatusBadRequest, ErrorResponse{Description: err.Error()})
+	}
+
+	endDate, err := util.StringQueryParam(ctx, "endDate", currentTime.Format(dateFormat))
+	fmt.Println(endDate)
+	if err != nil{
+		return "", "", ctx.JSON(http.StatusBadRequest, ErrorResponse{Description: err.Error()})
+	}
```

Comment:
I'm seeing a lot of these `verifyX` functions look pretty similar -- it'd probably be good to only have one place parsing out the dates and setting the defaults, rather than one in each file that does that. (The several you have are already inconsistent, as I commented elsewhere). It looks like one function doesn't even use a separate validation function. I'd put all the validation in its own file if it were me, probably. As long as they share a namespace, lowercased functions are still shared between files.

Reply:
I was thinking that keeping validation specific to each endpoint would be better if someone wanted to change parameters for a certain endpoint in the future?

Reply:
It's pretty easy to do that anyway, so it's probably worth changing for the don't-repeat-yourself benefits. For example, you could take a function much like this one, put it in a separate file, and call it something like `verifyDateParameters`. Then, in each endpoint that uses dates, you do `startDate, EndDate, err := verifyDateParameters(ctx)`. If you need to verify other parameters, you can write some other functions that handle those sorts of parameters -- say you wanted a user parameter but only on some endpoints, you could write a `verifyUserParameter` and then just have another line that does `user, err := verifyUserParameter(ctx)` that's only added to the relevant places.

Basically, you can end up with a situation where maybe the handlers have two or more lines of calls to verification functions, but those functions themselves encapsulate a lot of code that then doesn't have to be repeated (and since it isn't repeated, can't be subtly different, which can be a big source of bugs -- like my other comment noting that not every date parameter verification in your current code adds a day, so the end dates will behave inconsistently).